library/aesce.c

```diff
@@ -101,66 +101,157 @@ int mbedtls_aesce_has_support(void)
 #endif
 }
 
+MBEDTLS_OPTIMIZE_ALWAYS
 static uint8x16_t aesce_encrypt_block(uint8x16_t block,
                                       unsigned char *keys,
                                       int rounds)
 {
-    for (int i = 0; i < rounds - 1; i++) {
-        /* AES AddRoundKey, SubBytes, ShiftRows (in this order).
-         * AddRoundKey adds the round key for the previous round. */
-        block = vaeseq_u8(block, vld1q_u8(keys + i * 16));
-        /* AES mix columns */
-        block = vaesmcq_u8(block);
+    /* Assume either 10, 12 or 14 rounds */
```

Comment:
Unrolling loops is something the compiler should be very good at. I really don't like us doing this

Reply:
Maybe the compiler can take the hint if we pass an enum? (I haven't tried.)
```
enum number_of_rounds {TEN = 10; TWELVE = 12; FOURTEEN = 14;};
static uint8x16_t aesce_encrypt_block(uint8x16_t block,
                                      unsigned char *keys,
                                      enum number_of_rounds valid_rounds)
```
And make sure to only ever pass `TEN`, `TWELVE` or `FOURTEEN` when calling this function. (You can expect the compiler to be clever about `switch (rounds) {case 10:  valid_rounds = TEN; break; /*etc*/ default: return ERROR;} aesce_encrypt_block(...);`.)

Reply:
I think Gilles' suggestion is more complex than what we have. I've tidied-up a bit as per discussion with Tom, and strengthened the documentation warning about passing an invalid number of rounds.

As to whether to unroll at all, it gives us 56% for XTS (gcc -O2) and 30-40% on clang -O2. GCM benefits around 1% - 16%. So definitely worthwhile IMO.